File: 03-系统编程/01-网络编程/进程间通信/SCM-RIGHTS-fd传递/python/main.py
#!/usr/bin/env python3
"""SCM_RIGHTS / fd 传递 —— 可运行的自检 oracle。

README 里每一条"形式化断言"在这里都是一条真的断言。跑法：
    python3 main.py

模型见同目录 scm_model.py；C / Go 版本是同一套断言的移植。
"""

from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from scm_model import (  # noqa: E402
    EBADF, EINVAL, ETOOMANYREFS, MSG_CMSG_CLOEXEC, MSG_CTRUNC,
    SCM_MAX_FD, SCM_MAX_FD_OLD, FdTable, OpenFileDescription, ScmError,
    UnixSocket, cmsg_len, cmsg_space, fds_fit, fds_total_space,
)

from harness import check, open_file, pair, summary  # noqa: E402

# ================================================================ 场景 1
print("=== 1) 传的是 open file description 的引用（等价 dup），不是 fd 号 ===")
payload = b"HELLO-WORLD-0123456789"
ofd = open_file(payload)
sender = FdTable("sender", limit=64)
recv = FdTable("receiver", limit=64)

s_fd = sender.install(ofd)
# 接收方先占掉两个编号，好让它分配到的号与发送方不同 —— 编号是**每张表各自**的
recv.install(open_file(b"occupied-3"))
recv.install(open_file(b"occupied-4"))
print(f"  发送方 fd={s_fd} -> {sender.summary()}")
print(f"  发送方先读 5 字节: {ofd.read(5)!r}  → offset={ofd.offset}")

sock_a, sock_b = pair()
sent = sock_a.sendmsg(sender, b"!", [s_fd])
res = sock_b.recvmsg(recv, buf_size=16, ctrl_buf_size=64)
r_fd = res.fds[0]
print(f"  发送方 sendmsg(data={sent} B, fd={s_fd})")
print(f"  接收方 recvmsg → data={res.data!r} fds={res.fds} ({recv.summary()})")
print(f"  接收方读 5 字节: {recv.get(r_fd).ofd.read(5)!r}  → offset={ofd.offset}")

check("接收方分配到的 fd 号与发送方无关（编号是每张表各自的）", r_fd != s_fd,
      f"s={s_fd} r={r_fd}")
check("接收方读到的是 offset=5 之后的字节，说明共享同一个 OFD",
      recv.get(r_fd).ofd.offset == 10, f"offset={recv.get(r_fd).ofd.offset}")
check("两侧 fd 指向同一个 OFD 对象", recv.get(r_fd).ofd is ofd)

# 对照：按路径自己 open 会得到**另一个** OFD，offset 从 0 开始
fs = {"/tmp/demo.txt": payload}
other = open_file(fs["/tmp/demo.txt"])
after = other.read(5)
print(f"  对照：接收方按路径自己 open 再读 5 字节: {after!r}"
      f"  → offset={other.offset}")
check("按路径 re-open 得到独立 OFD，读出的内容完全不同", after != b"-WORL")
check("re-open 的 OFD 与传递得来的不是同一个对象", other is not ofd)

# 发送方 close 后接收方仍可用（引用计数：两张表各持一份引用）
before = ofd.refcount
sender.close(s_fd)
print(f"  发送方 close 掉 fd={s_fd} 前 refs={before}，"
      f"close 后 refs={ofd.refcount}，接收方仍可读 "
      f"{recv.get(r_fd).ofd.read(4)!r}")
check("两个进程各持一份引用：发送方 close 只掉 1", before == 2)
check("发送方 close 后接收方引用仍在", ofd.refcount == 1)

# ================================================================ 场景 2
print()
print(f"=== 2) 一次最多传的 fd 数：SCM_MAX_FD={SCM_MAX_FD}（<2.6.38 为 {SCM_MAX_FD_OLD}）===")
s2 = FdTable("s2", limit=4096)
r2 = FdTable("r2", limit=4096)
big = [s2.install(open_file(b"x")) for _ in range(SCM_MAX_FD)]
sock_a2, sock_b2 = pair()
sock_a2.sendmsg(s2, b"!", big)
res2 = sock_b2.recvmsg(r2, buf_size=8, ctrl_buf_size=fds_total_space(SCM_MAX_FD))
print(f"  传 {SCM_MAX_FD} 个 fd：发送成功，接收方装好 {len(res2.fds)} 个"
      f"（占控制缓冲 {fds_total_space(SCM_MAX_FD)} B）")
check(f"{SCM_MAX_FD} 个 fd 可以传", len(res2.fds) == SCM_MAX_FD,
      f"实际 {len(res2.fds)}")

extra = s2.install(open_file(b"x"))
try:
    sock_a2.sendmsg(s2, b"!", big + [extra])
    check(f"{SCM_MAX_FD + 1} 个 fd 应报 EINVAL", False, "居然没报错")
except ScmError as e:
    print(f"  传 {SCM_MAX_FD + 1} 个 fd：sendmsg 失败 → {e}")
    check(f"{SCM_MAX_FD + 1} 个 fd 报 EINVAL", e.errno_name == EINVAL, e.errno_name)

# ================================================================ 场景 3
print()
print("=== 3) 流式 socket 必须夹带 ≥1 字节真实数据；数据报不强制 ===")
s3 = FdTable("s3", limit=64)
f3 = s3.install(open_file(b"y"))
st_a, st_b = pair()
try:
    st_a.sendmsg(s3, b"", [f3])
    check("流式 socket 传 0 字节应报 EINVAL", False, "居然没报错")
except ScmError as e:
    print(f"  流式 socket sendmsg(data=0 B, fd) → {e}")
    check("流式 socket 传 0 字节报 EINVAL", e.errno_name == EINVAL, e.errno_name)

dg_a = UnixSocket("dgram")
dg_b = UnixSocket("dgram", peer=dg_a)
n = dg_a.sendmsg(s3, b"", [f3])
r3 = dg_b.recvmsg(FdTable("r3", 64), buf_size=8, ctrl_buf_size=64)
print(f"  数据报 socket sendmsg(data=0 B, fd) → 成功（{n} 字节），收到 {len(r3.fds)} 个 fd")
check("数据报 socket 允许 0 字节真实数据（但可移植代码仍应带 1 字节）",
      len(r3.fds) == 1)

# ================================================================ 场景 4
print()
print("=== 4) CMSG_SPACE vs CMSG_LEN：控制缓冲少算 4 字节就丢 fd ===")
print(f"  1 个 fd: cmsg_len = CMSG_LEN(4) = {cmsg_len(4)}"
      f"，但缓冲要留 CMSG_SPACE(4) = {cmsg_space(4)} B")
check("CMSG_SPACE 比 CMSG_LEN 多 4 字节对齐填充", cmsg_space(4) - cmsg_len(4) == 4)
for buf in (4, 20, 23, 24, 48):
    fit = fds_fit(buf, 1)
    print(f"  msg_controllen={buf:3d} → 能装 {fit} 个 fd")
check("按 sizeof(int)=4 预留缓冲 → 一个 fd 都收不到", fds_fit(4, 1) == 0)
check("按 CMSG_LEN=20 预留 → 仍然收不到（漏算填充）", fds_fit(20, 1) == 0)
check("按 CMSG_SPACE=24 预留 → 正好收到 1 个", fds_fit(24, 1) == 1)
print("  多 fd 时同样是 CMSG_SPACE(4n)：n=2 → "
      f"{fds_total_space(2)} B，n=3 → {fds_total_space(3)} B")
print(f"  ⚠ 对齐粒度：CMSG_SPACE(4)={cmsg_space(4)} 与 CMSG_SPACE(8)={cmsg_space(8)} "
      f"**相等** → 1 个 fd 和 2 个 fd 占的控制缓冲一样大")
check("1 个 fd 与 2 个 fd 所需的控制缓冲都是 24 B（对齐后合并）",
      cmsg_space(4) == cmsg_space(8) == 24)
check("因此按字节数反推「能装几个 fd」会算错：24 B 其实能装 2 个",
      fds_fit(24, 2) == 2 and fds_fit(24, 1) == 1)

# ================================================================ 场景 5
print()
print("=== 5) 控制缓冲过小 → MSG_CTRUNC，且多余 fd 在**接收方**被自动关闭 ===")
s5 = FdTable("s5", limit=64)
r5 = FdTable("r5", limit=64)
three = [s5.install(open_file(b"z")) for _ in range(3)]
ofds5 = [s5.get(fd).ofd for fd in three]
a5, b5 = pair()
a5.sendmsg(s5, b"!", three)
print(f"  发送后 refs（发送方表 1 份 + 在途 1 份）: {[o.refcount for o in ofds5]}")
res5 = b5.recvmsg(r5, buf_size=8, ctrl_buf_size=fds_total_space(2))   # 只够 2 个
print(f"  只给够 2 个 fd 的缓冲 → 收到 {len(res5.fds)} 个，"
      f"MSG_CTRUNC={bool(res5.msg_flags & MSG_CTRUNC)}，"
      f"被自动关闭 {res5.dropped_trunc} 个")
print(f"  接收后 refs: {[o.refcount for o in ofds5]}"
      f"  ← 前两个在接收方建了新引用(2)；第 3 个丢了在途引用且没建新引用(1)")
check("多余 fd 触发 MSG_CTRUNC", bool(res5.msg_flags & MSG_CTRUNC))
check("只装下 2 个 fd", len(res5.fds) == 2)
check("被关掉的第 3 个 fd 少了一份引用（接收方没留住它）",
      ofds5[2].refcount == 1, f"refs={ofds5[2].refcount}")
check("装下的前两个各多一份引用（接收方表持有）", ofds5[0].refcount == 2)

# 如果发送方事先撒手，被截断的那份引用就是**最后一份** → 对象当场释放
s5b = FdTable("s5b", limit=64)
r5b = FdTable("r5b", limit=64)
three_b = [s5b.install(open_file(b"z")) for _ in range(3)]
ofds5b = [s5b.get(fd).ofd for fd in three_b]
a5b, b5b = pair()
a5b.sendmsg(s5b, b"!", three_b)
for fd in three_b:
    s5b.close(fd)
print(f"  发送方发完就 close → 在途 refs: {[o.refcount for o in ofds5b]}"
      f"（只靠 socket 队列吊着）")
res5b = b5b.recvmsg(r5b, buf_size=8, ctrl_buf_size=fds_total_space(2))
print(f"  控制缓冲给 {fds_total_space(2)} B（够 2 个）→ 收到 {len(res5b.fds)} 个，"
      f"被关闭的第 3 个 refs={ofds5b[2].refcount} ← 归零即释放")
check("没有任何一方再持有 → 被截断的 fd 引用计数归零（对象释放）",
      ofds5b[2].refcount == 0, f"refs={ofds5b[2].refcount}")
check("留下来的那两个各被接收方表持有一份", 
      all(o.refcount == 1 for o in ofds5b[:2]))

# ================================================================ 场景 6
print()
print("=== 6) 接收方 RLIMIT_NOFILE 不足 → 同样自动关闭，且不报错 ===")
s6 = FdTable("s6", limit=64)
r6 = FdTable("r6", limit=3)          # 只剩很少槽位
r6.install(open_file(b"occupied"))   # 先占掉一个
ten = [s6.install(open_file(b"q")) for _ in range(10)]
ofds6 = [s6.get(fd).ofd for fd in ten]
a6, b6 = pair()
a6.sendmsg(s6, b"!", ten)
res6 = b6.recvmsg(r6, buf_size=8, ctrl_buf_size=fds_total_space(10))
print(f"  RLIMIT_NOFILE=3（已用 1）→ 装好 {len(res6.fds)} 个，"
      f"因限流自动关闭 {res6.dropped_rlimit} 个，recvmsg 本身**没有**报错")
print(f"  接收方表: {r6.summary()}")
check("recvmsg 不因 RLIMIT_NOFILE 报错（静默丢弃）", True)
check("装下的数量不超过 RLIMIT_NOFILE 剩余槽位",
      len(res6.fds) <= r6.limit - 1, f"{len(res6.fds)}")
check("超出限流的 fd 被自动关闭", res6.dropped_rlimit >= 1,
      f"dropped={res6.dropped_rlimit}")

# ================================================================ 场景 7
print()
print("=== 7) 在途（in-flight）fd 记账：Linux 4.5 起会报 ETOOMANYREFS ===")
s7 = FdTable("s7", limit=8)
a7, b7 = pair()                      # b7 一直不 recvmsg，制造"在途"
err7 = None
table_always_empty = True
for i in range(1, 13):
    fd = s7.install(open_file(b"f"))
    try:
        a7.sendmsg(s7, b"!", [fd])
    except ScmError as e:
        err7 = e
        s7.close(fd)
        print(f"  第 {i} 次 sendmsg（在途已 {a7.in_flight()}）→ {e}")
        break
    s7.close(fd)                     # 关键：发完立刻 close，旧内核靠这招绕过记账
    if s7.size() != 0 or a7.in_flight() != i:
        table_always_empty = False
    print(f"  第 {i} 次 sendmsg 成功，发送方 close 掉本地 fd，"
          f"在途={a7.in_flight()}，发送方表={s7.size()} 项")
check("在途 fd 超限报 ETOOMANYREFS",
      err7 is not None and err7.errno_name == ETOOMANYREFS,
      err7.errno_name if err7 else "没报错")
check("旧内核漏洞的成因：发送方 fd 表恒空而在途量持续累积到 RLIMIT_NOFILE",
      table_always_empty and a7.in_flight() == 8)
print(f"  −> Linux 4.5 之前这里能无限发下去（发送方表恒空，RLIMIT_NOFILE 形同虚设）；"
      f"4.5 起按在途量记账并报 ETOOMANYREFS，除非有 CAP_SYS_RESOURCE")
# 对端把消息全接走之后在途量才清零（一条消息一次 recvmsg：控制数据是屏障）
r7 = FdTable("r7", 4096)
while b7.recvmsg(r7, buf_size=8, ctrl_buf_size=64) is not None:
    pass
print(f"  对端逐条 recvmsg 接走后在途={a7.in_flight()}")
check("对端接走后在途量清零", a7.in_flight() == 0)

# ================================================================ 场景 8
print()
print("=== 8) 控制数据是屏障：man7 unix(7) 的 4 / 1 / 4 字节例子 ===")
s8 = FdTable("s8", limit=64)
f8 = s8.install(open_file(b"m"))
a8, b8 = pair()
a8.sendmsg(s8, b"AAAA", [])          # (1) 4 字节，无控制数据
a8.sendmsg(s8, b"B", [f8])           # (2) 1 字节 + 控制数据
a8.sendmsg(s8, b"CCCC", [])          # (3) 4 字节，无控制数据
r8 = FdTable("r8", limit=64)
first = b8.recvmsg(r8, buf_size=20, ctrl_buf_size=64)
second = b8.recvmsg(r8, buf_size=20, ctrl_buf_size=64)
third = b8.recvmsg(r8, buf_size=20, ctrl_buf_size=64)
third_data = third.data if third else None
print(f"  第 1 次 recvmsg(buf=20) → {first.data!r} + {len(first.fds)} 个 fd")
print(f"  第 2 次 recvmsg(buf=20) → {second.data!r}")
print(f"  第 3 次 recvmsg(buf=20) → {third_data!r}（队列已空则 None）")
check("第 1 次拿到 5 字节（4+1）并在同一调用里拿到 fd",
      first.data == b"AAAAB" and len(first.fds) == 1, f"{first.data!r}")
check("屏障挡住了后面那 4 字节", second.data == b"CCCC", f"{second.data!r}")

# ================================================================ 场景 9
print()
print("=== 9) MSG_CMSG_CLOEXEC（Linux 2.6.23）：消除 exec 泄漏窗口 ===")
s9 = FdTable("s9", limit=64)
a9, b9 = pair()
r9a = FdTable("r9a", limit=64)
r9b = FdTable("r9b", limit=64)
f9 = s9.install(open_file(b"c"))
a9.sendmsg(s9, b"!", [f9])
plain = b9.recvmsg(r9a, buf_size=8, ctrl_buf_size=64)
a9.sendmsg(s9, b"!", [f9])
withflag = b9.recvmsg(r9b, buf_size=8, ctrl_buf_size=64, flags=MSG_CMSG_CLOEXEC)
print(f"  不带 flag → FD_CLOEXEC={r9a.get(plain.fds[0]).cloexec}")
print(f"  带 MSG_CMSG_CLOEXEC → FD_CLOEXEC={r9b.get(withflag.fds[0]).cloexec}")
check("不带 flag 时收到的新 fd 不带 CLOEXEC（要自己 fcntl，存在竞态）",
      r9a.get(plain.fds[0]).cloexec is False)
check("MSG_CMSG_CLOEXEC 原子地设好 CLOEXEC", r9b.get(withflag.fds[0]).cloexec is True)

# ================================================================ 场景 10
print()
print("=== 10) 发送一个不存在的 fd → EBADF ===")
s10 = FdTable("s10", limit=64)
a10, _ = pair()
try:
    a10.sendmsg(s10, b"!", [42])
    check("发送无效 fd 应报 EBADF", False, "居然没报错")
except ScmError as e:
    print(f"  sendmsg(fd=42，本进程没有) → {e}")
    check("发送无效 fd 报 EBADF", e.errno_name == EBADF, e.errno_name)

sys.exit(summary())

File: 03-系统编程/01-网络编程/进程间通信/SCM-RIGHTS-fd传递/go/main.go
// main.go — SCM_RIGHTS / fd 传递 demo（Go）。
//
// 真实系统调用路径 —— Go 标准库把送控制消息的活拆成了三个函数：
//   syscall.UnixRights(fds...)          → 造出 SCM_RIGHTS 的 oob 字节串
//   syscall.Sendmsg(sock, data, oob, …) → 发（**必须**同时给 >=1 字节 data）
//   syscall.Recvmsg(...) + syscall.ParseSocketControlMessage + ParseUnixRights
// 注意 UnixRights/ParseUnixRights 按"int 为 4 字节"编码，只在 Linux 正确。
//
// 编译： cd go && go build -o scm_demo .
//        或直接 go run main.go scm_model.go
package main

import (
	"fmt"
	"os"
	"syscall"
)

var (
	nOk   int
	nFail int
)

func check(name string, cond bool, format string, args ...any) {
	if cond {
		nOk++
		fmt.Printf("  [ok]   %s\n", name)
		return
	}
	nFail++
	fmt.Printf("  [FAIL] %s  %s\n", name, fmt.Sprintf(format, args...))
}

// ---------------------------------------------------------------- 真实路径

// realSend 用真实 sendmsg 送 len(data) 字节 + n 个 fd。
func realSend(sock int, data []byte, fds []int, errnoOnly bool) error {
	oob := syscall.UnixRights(fds...)
	if errnoOnly {
		oob = nil
	}
	return syscall.Sendmsg(sock, data, oob, nil, 0)
}

// realRecv 收一次；返回数据、fd 列表、msg_flags。
func realRecv(sock int, bufSize, ctrlBufSize int, flags int) ([]byte, []int, int, error) {
	buf := make([]byte, bufSize)
	oob := make([]byte, ctrlBufSize)
	n, oobn, recvFlags, _, err := syscall.Recvmsg(sock, buf, oob, flags)
	if err != nil {
		return nil, nil, 0, err
	}
	var fds []int
	if oobn > 0 {
		msgs, perr := syscall.ParseSocketControlMessage(oob[:oobn])
		if perr != nil {
			return nil, nil, recvFlags, perr
		}
		for i := range msgs {
			got, perr := syscall.ParseUnixRights(&msgs[i])
			if perr == nil {
				fds = append(fds, got...)
			}
		}
	}
	return buf[:n], fds, recvFlags, nil
}

// scenarioIdentity 证明"传的是 open file description 的引用"。
func scenarioLimits() {
	fmt.Printf("\n=== 2) SCM_MAX_FD=%d（<2.6.38 为 %d）；3) 必须带真实数据；7) EBADF ===\n",
		scmMaxFd, scmMaxFdOld)
	pair, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
	if err != nil {
		check("socketpair", false, "%v", err)
		return
	}
	defer syscall.Close(pair[0])
	defer syscall.Close(pair[1])

	many := make([]int, 0, scmMaxFd+1)
	for i := 0; i < scmMaxFd+1; i++ {
		dup, derr := syscall.Dup(0)
		if derr != nil {
			check("准备 dup 的 fd", false, "%v", derr)
			return
		}
		many = append(many, dup)
	}
	if err := realSend(pair[0], []byte("!"), many[:scmMaxFd], false); err != nil {
		check("传 253 个 fd 成功", false, "%v", err)
	} else {
		_, fds, _, rerr := realRecv(pair[1], 8, fdsSpace(scmMaxFd), 0)
		fmt.Printf("  传 %d 个 fd：发送成功，接收方装好 %d 个（控制缓冲 %d B）\n",
			scmMaxFd, len(fds), fdsSpace(scmMaxFd))
		check("253 个 fd 可以一次性传完", rerr == nil && len(fds) == scmMaxFd,
			"got=%d err=%v", len(fds), rerr)
		for _, fd := range fds {
			syscall.Close(fd)
		}
	}
	err = realSend(pair[0], []byte("!"), many, false)
	fmt.Printf("  传 %d 个 fd：sendmsg → %v\n", len(many), err)
	check("254 个 fd 报 EINVAL", err == syscall.EINVAL, "err=%v", err)

	err = realSend(pair[0], nil, many[:1], false)
	fmt.Printf("  流式 socket sendmsg(data=0 B, fd) → %v\n", err)
	check("流式 socket 传 0 字节报 EINVAL", err == syscall.EINVAL, "err=%v", err)

	err = realSend(pair[0], []byte("!"), []int{99999}, false)
	fmt.Printf("  sendmsg(fd=99999，本进程没有) → %v\n", err)
	check("发送无效 fd 报 EBADF", err == syscall.EBADF, "err=%v", err)
}

// scenarioTruncation 控制缓冲过小 → MSG_CTRUNC，多余的 fd 由内核在接收方关闭。
func scenarioTruncation() {
	fmt.Println("\n=== 4) CMSG_SPACE vs CMSG_LEN；缓冲过小 → MSG_CTRUNC ===")
	fmt.Printf("  1 个 fd: CMSG_LEN(4)=%d（写进 cmsg_len），CMSG_SPACE(4)=%d（要预留的缓冲）\n",
		cmsgLen(4), cmsgSpace(4))
	check("CMSG_SPACE 比 CMSG_LEN 多一个对齐填充", cmsgSpace(4)-cmsgLen(4) == 4,
		"%d vs %d", cmsgSpace(4), cmsgLen(4))
	fmt.Printf("  fdsFit(4,1)=%d  fdsFit(%d,1)=%d  fdsFit(%d,1)=%d\n",
		fdsFit(4, 1), cmsgLen(4), fdsFit(cmsgLen(4), 1), cmsgSpace(4), fdsFit(cmsgSpace(4), 1))
	check("按 sizeof(int)=4 预留 → 一个都收不到", fdsFit(4, 1) == 0, "")
	check("按 CMSG_LEN=20 预留 → 仍然收不到", fdsFit(cmsgLen(4), 1) == 0, "")
	check("按 CMSG_SPACE=24 预留 → 正好收到 1 个", fdsFit(cmsgSpace(4), 1) == 1, "")
	fmt.Printf("  ⚠ 对齐粒度：CMSG_SPACE(4)=%d 与 CMSG_SPACE(8)=%d 相等 → "+
		"1 个和 2 个 fd 占的缓冲一样大\n", cmsgSpace(4), cmsgSpace(8))
	check("按 CMSG_SPACE(4*2)=24 预留其实能装 2 个", fdsFit(cmsgSpace(4), 2) == 2, "")

	pair, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
	if err != nil {
		check("socketpair", false, "%v", err)
		return
	}
	defer syscall.Close(pair[0])
	defer syscall.Close(pair[1])
	three := make([]int, 0, 3)
	for i := 0; i < 3; i++ {
		dup, _ := syscall.Dup(0)
		three = append(three, dup)
	}
	if err := realSend(pair[0], []byte("!"), three, false); err != nil {
		check("发送 3 个 fd", false, "%v", err)
		return
	}
	_, fds, recvFlags, err := realRecv(pair[1], 8, cmsgSpace(4), 0)
	fmt.Printf("  发送 3 个 fd、控制缓冲只给 %d B → 收到 %d 个，MSG_CTRUNC=%v\n",
		cmsgSpace(4), len(fds), recvFlags&syscall.MSG_CTRUNC != 0)
	check("多余 fd 触发 MSG_CTRUNC", recvFlags&syscall.MSG_CTRUNC != 0, "flags=%d", recvFlags)
	check("只装下 2 个 fd", len(fds) == 2, "got=%d err=%v", len(fds), err)
	fmt.Println("  ← 被截断的那个 fd 由内核在接收进程里自动关闭（不会退回发送方）")
	for _, fd := range fds {
		syscall.Close(fd)
	}
	for _, fd := range three {
		syscall.Close(fd)
	}
}

// scenarioBarrier 屏障 + MSG_CMSG_CLOEXEC。
func scenarioBarrier() {
	fmt.Println("\n=== 5) 控制数据是屏障；6) MSG_CMSG_CLOEXEC ===")
	pair, err := syscall.Socketpair(syscall.AF_UNIX, syscall.SOCK_STREAM, 0)
	if err != nil {
		check("socketpair", false, "%v", err)
		return
	}
	defer syscall.Close(pair[0])
	defer syscall.Close(pair[1])
	fd, _ := syscall.Dup(0)
	defer syscall.Close(fd)

	syscall.Write(pair[0], []byte("AAAA"))
	if err := realSend(pair[0], []byte("B"), []int{fd}, false); err != nil {
		check("sendmsg 中段", false, "%v", err)
		return
	}
	syscall.Write(pair[0], []byte("CCCC"))

	data, fds, _, err := realRecv(pair[1], 20, 256, syscall.MSG_CMSG_CLOEXEC)
	if err != nil {
		check("第 1 次 recvmsg", false, "%v", err)
		return
	}
	fmt.Printf("  第 1 次 recvmsg(buf=20) → %q + %d 个 fd\n", data, len(fds))
	check("一次拿到 5 字节（前 4 + 带控制数据的 1）",
		string(data) == "AAAAB" && len(fds) == 1, "data=%q fds=%v", data, fds)
	if len(fds) == 1 {
		flags, _, ferr := syscall.Syscall(syscall.SYS_FCNTL, uintptr(fds[0]), syscall.F_GETFD, 0)
		fmt.Printf("  带 MSG_CMSG_CLOEXEC 收到的 fd：FD_CLOEXEC=%v\n",
			ferr == 0 && flags&syscall.FD_CLOEXEC != 0)
		check("MSG_CMSG_CLOEXEC 原子地设好 FD_CLOEXEC",
			ferr == 0 && flags&syscall.FD_CLOEXEC != 0, "flags=%d err=%v", flags, ferr)
		syscall.Close(fds[0])
	}
	data2, _, _, err := realRecv(pair[1], 20, 256, 0)
	if err != nil {
		check("第 2 次 recvmsg", false, "%v", err)
		return
	}
	fmt.Printf("  第 2 次 recvmsg(buf=20) → %q（屏障挡住了这 4 字节）\n", data2)
	check("屏障挡住后面那段：第 2 次才拿到 CCCC", string(data2) == "CCCC", "got=%q", data2)
}

// scenarioModel 纯模型部分：在途 fd 记账 + RLIMIT_NOFILE（真实内核上很难稳定复现）。
func scenarioModel() {
	fmt.Println("\n=== 8) 纯模型：在途 fd 记账（ETOOMANYREFS）与接收方 RLIMIT ===")
	sender := newFdTable("sender", 8)
	sock, peer := newPair("stream")
	var lastErr error
	alwaysEmpty := true
	for i := 1; i <= 12; i++ {
		ofd := newOFD("file", []byte("f"))
		fd, _ := sender.install(ofd, false)
		if _, err := sock.sendmsg(sender, []byte("!"), []int{fd}, false); err != nil {
			lastErr = err
			sender.close(fd)
			break
		}
		sender.close(fd) // 发完立刻 close：旧内核靠这招绕过 RLIMIT_NOFILE
		if sender.size() != 0 || peer.pendingFds() != i {
			alwaysEmpty = false
		}
	}
	fmt.Printf("  发送方 fd 表恒空、在途累积到 %d 后 → %v\n", peer.pendingFds(), lastErr)
	check("在途 fd 超限报 ETOOMANYREFS",
		lastErr != nil && lastErr.(*scmError).errno == "ETOOMANYREFS", "err=%v", lastErr)
	check("旧内核漏洞的成因：本进程表恒空而在途持续累积", alwaysEmpty, "")

	recv := newFdTable("receiver", 3)
	recv.install(newOFD("file", []byte("occupied")), false)
	out := peer.recvmsg(recv, 8, fdsSpace(12), 0)
	fmt.Printf("  接收方 RLIMIT_NOFILE=3（已用 1，在途 8）→ 装好 %d 个，"
		"因限流自动关闭 %d 个，recvmsg 本身没报错\n", len(out.fds), out.droppedRlim)
	check("装下的数量不超过剩余槽位", len(out.fds) <= 2, "got=%d", len(out.fds))
	check("超限的 fd 被自动关闭（且不置 MSG_CTRUNC —— man 页未规定）",
		out.droppedRlim >= 1 && out.msgFlags&msgCtrunc == 0, "drop=%d", out.droppedRlim)
}

func main() {
	fmt.Println("SCM_RIGHTS / fd 传递 demo —— Go（syscall.UnixRights 真实路径）\n")
	if syscall.Gettid() < 0 {
		fmt.Println("（当前平台不支持）")
		return
	}
	scenarioIdentity()
	scenarioLimits()
	scenarioTruncation()
	scenarioBarrier()
	scenarioModel()
	fmt.Printf("\n===== 断言结果: %d/%d 通过，%d 失败 =====\n", nOk, nOk+nFail, nFail)
	if nFail > 0 {
		os.Exit(1)
	}
}
